includes/Utilities/Plugin_Request_Utility.php

```diff
@@ -182,6 +182,91 @@ public static function get_files_to_ignore() {
 		return $files_to_ignore;
 	}
 
+	/**
+	 * Returns the plugin basename after downloading and installing the plugin.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param string $plugin_url The URL of the plugin to download.
+	 * @return string The plugin basename after downloading and installing the plugin.
+	 *
+	 * @throws Exception Thrown if an invalid URL given or zip could be extracted properly.
+	 */
+	public static function download_plugin( $plugin_url ) {
+		$response = wp_remote_get( $plugin_url );
+
+		if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
+			throw new Exception(
+				__( 'Downloading the zip file failed.', 'plugin-check' )
+			);
+		}
+
+		// Prevents URL from wporg.
+		if ( false !== strpos( $plugin_url, '#wporgapi:' ) ) {
+			$plugin_info_url = substr( $plugin_url, strpos( $plugin_url, '#wporgapi:' ) );
+			$plugin_info_url = str_replace( '#wporgapi:', '', $plugin_info_url );
+			$plugin_url      = substr( $plugin_url, 0, strpos( $plugin_url, '#' ) );
+		}
+
+		$basename = basename( $plugin_url );
+
+		// Create the name of the file and the declare the directory and path.
+		$plugin_check_dir = self::get_upload_dir();
+
+		$file_path    = $plugin_check_dir . $basename;
+		$file_process = fopen( $file_path, 'w' );
+		fwrite( $file_process, $response['body'] );
+		fclose( $file_process );
+
+		$temp_dir = $plugin_check_dir . strtotime( 'now' ) . '/';
+
+		// Unzip file.
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+
+		WP_Filesystem();
+
+		$unzip_file = unzip_file( $file_path, $temp_dir );
+
+		if ( true !== $unzip_file ) {
+			throw new Exception( $unzip_file->get_error_message() );
+		}
+
+		// Remove zip file.
+		unlink( $file_path );
+
+		if ( ! empty( $plugin_info_url ) ) {
```

Comment:
Would be good to validate this URL before using, maybe also with `filter_var( $plugin_info_url, FILTER_VALIDATE_URL )`